x-pack/plugins/apm/server/lib/service_map/get_service_anomalies.ts

```diff
@@ -0,0 +1,158 @@
+  const apmMlJobCategories = apmMlJobs
+    .map((job) => getApmMlJobCategory(job, serviceNames))
+    .filter(
+      (apmJobCategory) => apmJobCategory !== undefined
+    ) as ApmMlJobCategory[];
+  const apmJobIds = apmMlJobs.map((job) => job.job_id);
+  const rangeQuery = { range: rangeFilter(start, end, 'timestamp') };
+  const params = {
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { term: { result_type: 'bucket' } },
+            {
+              terms: {
+                job_id: apmJobIds,
+              },
+            },
+            rangeQuery,
+          ],
+        },
+      },
+      aggs: {
+        jobs: {
+          terms: { field: 'job_id', size: apmJobIds.length },
+          aggs: {
+            top_score_hits: {
+              top_hits: {
+                sort: [{ anomaly_score: { order: 'desc' as const } }],
+                _source: ['anomaly_score', 'timestamp'],
+                size: 1,
+              },
+            },
+          },
+        },
+      },
+    },
+  };
+
+  const response = await ml.mlSystem.mlAnomalySearch(params);
+  const anomalyScores: Array<{
+    jobId: string;
+    anomalyScore: number;
+    timestamp: number;
+  }> = response.aggregations.jobs.buckets.map((jobBucket: any) => {
+    const jobId = jobBucket.key;
+    const bucketSource = jobBucket.top_score_hits.hits.hits?.[0]?._source;
+    return {
+      jobId,
+      anomalyScore: bucketSource.anomaly_score,
+      timestamp: bucketSource.timestamp,
+    };
+  });
+  const anomalyModelValuePromises = anomalyScores.map(
+    ({ jobId, timestamp }) => {
+      return (async () => {
+        try {
+          const modelPlotResponse = await ml.mlSystem.mlAnomalySearch({
+            body: {
+              size: 1,
+              query: {
+                bool: {
+                  filter: [
+                    { term: { result_type: 'model_plot' } },
+                    { term: { job_id: jobId } },
+                    { term: { timestamp } },
+                  ],
+                },
+              },
+              _source: ['actual', 'model_median'],
+            },
+          });
+          if (modelPlotResponse.hits.hits.length === 0) {
+            return;
+          }
+          const { actual, model_median } = modelPlotResponse.hits.hits[0]
+            ._source as {
+            actual: number;
+            model_median: number;
+          };
+          return { jobId, actual, model_median };
+        } catch (error) {
+          return;
+        }
+      })();
+    }
+  );
+
+  const anomalyModelValues = filterNonNullable(
+    await Promise.all(anomalyModelValuePromises)
+  );
+
+  return leftJoin(
+    apmMlJobCategories,
+    'jobId',
+    leftJoin(anomalyScores, 'jobId', anomalyModelValues)
+  );
+}
+
+function filterNonNullable<T>(array: T[]) {
+  return array.filter(
+    (element) => element !== undefined || element !== null
+  ) as Array<NonNullable<T>>;
+}
```

Comment:
Instead of overriding the compiler with `as` it's normally better to do something like:

```ts
function filterNonNullable<T>(value: T | null | undefined): value is T {
    return value !== null && value !== undefined;
}

// ...

return array.filter(filterNonNullable)
```

It uses the `is` syntax which is not bulletproof but better than `as` imo.

Reply:
Thanks for the tip. What about in-lining it since it's only used in 1 place?
```
  const anomalyModelValues = (
    await Promise.all(anomalyModelValuePromises)
  ).filter(<T>(value: T | undefined): value is T => value !== undefined);
```

Reply:
I find it more readable in a separate function but no big deal.